File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<int,int>hash;

        int i=0;
        int j=0;
         
        int maxLen=INT_MIN;
        while(j<s.length()){
            while(hash[s[j]]>0){
                hash[s[i]]--;
                i++;
            }

            maxLen=max(maxLen,j-i+1);
            hash[s[j]]++;
            j++;
        }

        return (maxLen==INT_MIN?0:maxLen);
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer>hash=new HashMap<>();

        int i=0;
        int j=0;
        int n=s.length();
        int maxLen=0;

        while(j<n){
            char ch=s.charAt(j);
            while(hash.containsKey(ch)){
               hash.remove(s.charAt(i));
               i++;
            }

            hash.put(ch,j);

            maxLen=Math.max(maxLen,j-i+1);

            j++;
        }


        return maxLen;
    }
}